proc_macros/src/lib.rs

```diff
@@ -0,0 +1,350 @@
+use std::collections::HashMap;
+
+use inflector::Inflector;
+use proc_macro::TokenStream;
+use quote::{format_ident, quote};
+use syn::{
+    bracketed,
+    parse::{Parse, ParseStream},
+    parse_macro_input, Expr, ExprArray, Ident, Token,
+};
+
+fn expr_to_ident(expr: Expr) -> Ident {
+    match expr {
+        Expr::Path(expr) => expr.path.get_ident().unwrap().clone(),
+        _ => panic!("Expected Ident"),
+    }
+}
+
+fn parse_idents_array(input: ParseStream) -> syn::Result<Vec<Ident>> {
+    Ok(input
+        .parse::<ExprArray>()?
+        .elems
+        .into_iter()
+        .map(expr_to_ident)
+        .collect())
+}
+
+struct FixedMapArgs {
+    name: Ident,
+    variants: Vec<Ident>,
+}
+
+impl Parse for FixedMapArgs {
+    fn parse(input: ParseStream) -> syn::Result<Self> {
+        let mut name: Option<Ident> = Default::default();
+        let mut variants: Option<Vec<Ident>> = Default::default();
+        while !input.is_empty() {
+            let key: Ident = input.parse()?;
+            input.parse::<Token![=>]>()?;
+            match &*key.to_string() {
+                "name" => {
+                    name = Some(input.parse::<Ident>()?);
+                }
+                "variants" => {
```

Comment:
Whoa, interesting. I guess I haven't seen this kind of stuff in a proc-macro before.

Reply:
Ya I was sort of "winging it" off of picturing the `macro! { key => value, other_key => other_value }` "invocation style"